> I have had a chance to test on darwin, here are the problems.
> 
>     * `which: no binarycreator` I am not sure where binarycreator is supposed to be found. I think it's just a matter of doing a substitution, but don't know where the package bin is supposed to be

@happysalada, `binarycreator` is in `qtinstaller`, it raises a warning because the strict `:=` is used instead of the lazy `=` in the `Makefile`, but is only needed by the `installer` target which is not used at this point. As documented in `docs/build-osx.rst`, `installer` is used to build the `.dmg` but I _guess_ that's not Nixpkgs's way to install on `Darwin`.
> 
>     * `iproute2` and `iptables` are packages specific to linux, so their substitution needs to be conditional on linux.

I've guarded uses of `bitmask-root` behind a `stdenv.isLinux` (AFAIU, `pkg/helper` is used on `Windows` and `Darwin`). Thanks!

>     * missing debug libs
> ```
> clang-7: error: no such file or directory: '/nix/store/xigx8vghkn27ykhm4gw7dyfkvvfw1kw0-qtbase-5.12.10/lib
> /libQt5Widgets_debug.dylib'
> clang-7: error: no such file or directory: '/nix/store/i3ff2saa275pvcrz57gr4q0dz371czm5-qtdeclarative-5.12
> .10/lib/libQt5Quick_debug.dylib'
> clang-7: error: no such file or directory: '/nix/store/xigx8vghkn27ykhm4gw7dyfkvvfw1kw0-qtbase-5.12.10/lib
> /libQt5Gui_debug.dylib'
> clang-7: error: no such file or directory: '/nix/store/i3ff2saa275pvcrz57gr4q0dz371czm5-qtdeclarative-5.12
> .10/lib/libQt5Qml_debug.dylib'
> clang-7: error: no such file or directory: '/nix/store/xigx8vghkn27ykhm4gw7dyfkvvfw1kw0-qtbase-5.12.10/lib
> /libQt5Network_debug.dylib'
> clang-7: error: no such file or directory: '/nix/store/xigx8vghkn27ykhm4gw7dyfkvvfw1kw0-qtbase-5.12.10/lib
> /libQt5Core_debug.dylib'
> ```
> 
> I'm not sure how to include those debug libs, I've tried adding `qtbase.dev` and `qtdeclarative.dev` in buildInputs but it didn't change anything.

Hm, I've no idea, sometimes it works better in `nativeBuildInputs`.

>     * There are some missing symbols
> ```
> Undefined symbols for architecture x86_64:
>   "_CFArrayGetCount", referenced from:
>       _crypto/x509/internal/macos.x509_CFArrayGetCount_trampoline in libgoshim.a(go.o)
>   "_CFArrayGetValueAtIndex", referenced from:
>       _crypto/x509/internal/macos.x509_CFArrayGetValueAtIndex_trampoline in libgoshim.a(go.o)
>   "_CFDataGetBytePtr", referenced from:
>       _crypto/x509/internal/macos.x509_CFDataGetBytePtr_trampoline in libgoshim.a(go.o)
>   "_CFDataGetLength", referenced from:
>       _crypto/x509/internal/macos.x509_CFDataGetLength_trampoline in libgoshim.a(go.o)
>   "_CFDictionaryGetValueIfPresent", referenced from:
>       _crypto/x509/internal/macos.x509_CFDictionaryGetValueIfPresent_trampoline in libgoshim.a(go.o)
>   "_CFEqual", referenced from:
>       _crypto/x509/internal/macos.x509_CFEqual_trampoline in libgoshim.a(go.o)
>   "_CFNumberGetValue", referenced from:
>       _crypto/x509/internal/macos.x509_CFNumberGetValue_trampoline in libgoshim.a(go.o)
>   "_CFRelease", referenced from:
>       _crypto/x509/internal/macos.x509_CFRelease_trampoline in libgoshim.a(go.o)
>   "_CFStringCreateWithBytes", referenced from:
>       _crypto/x509/internal/macos.x509_CFStringCreateWithBytes_trampoline in libgoshim.a(go.o)
> ld: symbol(s) not found for architecture x86_64
> ```
> those are in the CoreFoundation framework, it seems though that a new flag needs to passed to clang. Is the right way to set the `NIX_LDFLAGS` variables ? I'm not sure. The flag I want to pass is `-framework CoreFoundation`.

I would try to patch `bitmask.pro` where there's already a `-framework` flag, but this seems very strange that such flag is missing.
But `git grep -i corefoundation` suggests rather to add `CoreFoundation` to `buildInputs`. i've done that and also added `Security` (which appears in `bitmask.pro`).

> I have time to test this everyday if needed. I realise though that you might want to split darwin support in a separate PR, if you would rather do that, I understand.

Since I am unable to develop on Darwin, it would indeed be better for me to pass the relay and let you shepherd `Darwin` support on a following PR. In the meantime I've added what I can to improve the state of this package on `Darwin` based on our discussions here.

--- a/pkgs/tools/networking/bitmask-vpn/default.nix
+++ b/pkgs/tools/networking/bitmask-vpn/default.nix
@@ -1,0 +1,168 @@
+{ lib
+, stdenv
+, fetchFromGitLab
+, mkDerivation
+, buildGoModule
+, wrapQtAppsHook
+, python3Packages
+, pkg-config
+, openvpn
+, cmake
+, qmake
+, which
+, iproute2
+, iptables
+, procps
+, qmltermwidget
+, qtbase
+, qtdeclarative
+, qtinstaller
+, qtquickcontrols
+, qtquickcontrols2
+, qttools
+, CoreFoundation
+, Security
+, provider ? "riseup"
+}:
+let
+  version = "0.21.6";
+
+  src = fetchFromGitLab {
+    domain = "0xacab.org";
+    owner = "leap";
+    repo = "bitmask-vpn";
+    rev = version;
+    sha256 = "sha256-LMz+ZgQVFGujoLA8rlyZ3VnW/NSlPipD5KwCe+cFtnY=";
+  };
+
+  # bitmask-root is only used on GNU/Linux
+  # and may one day be replaced by pkg/helper
+  bitmask-root = mkDerivation {
+    inherit src version;
+    sourceRoot = "source/helpers";
+    pname = "bitmask-root";
+    nativeBuildInputs = [ python3Packages.wrapPython ];
+    postPatch = ''
+      substituteInPlace bitmask-root \
+        --replace 'swhich("ip")' '"${iproute2}/bin/ip"' \
+        --replace 'swhich("iptables")' '"${iptables}/bin/iptables"' \
+        --replace 'swhich("ip6tables")' '"${iptables}/bin/ip6tables"' \
+        --replace 'swhich("sysctl")' '"${procps}/bin/sysctl"' \
+        --replace /usr/sbin/openvpn ${openvpn}/bin/openvpn
+      substituteInPlace se.leap.bitmask.policy \
+        --replace /usr/sbin/bitmask-root $out/bin/bitmask-root
+    '';
+    installPhase = ''
+      runHook preInstall
+
+      install -m 755 -D -t $out/bin bitmask-root
+      install -m 444 -D -t $out/share/polkit-1/actions se.leap.bitmask.policy
+      wrapPythonPrograms
+
+      runHook postInstall
+    '';
+  };
+in
+
+buildGoModule rec {
+  inherit src version;
+  pname = "${provider}-vpn";
+  vendorSha256 = null;
+
+  postPatch = ''
+    substituteInPlace pkg/pickle/helpers.go \
+      --replace /usr/share $out/share
+
+    # Using $PROVIDER is not working,
+    # thus replacing directly into the vendor.conf
+    substituteInPlace providers/vendor.conf \
+      --replace "provider = riseup" "provider = ${provider}"
+
+    patchShebangs gui/build.sh
+    wrapPythonProgramsIn branding/scripts
+  '' + lib.optionalString stdenv.isLinux ''
+    substituteInPlace pkg/helper/linux.go \
+      --replace /usr/sbin/openvpn ${openvpn}/bin/openvpn
+    substituteInPlace pkg/vpn/launcher_linux.go \
+      --replace /usr/sbin/openvpn ${openvpn}/bin/openvpn \
+      --replace /usr/sbin/bitmask-root ${bitmask-root}/bin/bitmask-root \
+      --replace /usr/bin/lxpolkit /run/wrappers/bin/polkit-agent-helper-1 \
+      --replace '"polkit-gnome-authentication-agent-1",' '"polkit-gnome-authentication-agent-1","polkitd",'
+  '';
+
+  nativeBuildInputs = [
+    cmake
+    pkg-config
+    python3Packages.wrapPython
+    qmake
+    qtquickcontrols
+    qtquickcontrols2
+    qttools
+    which
+    wrapQtAppsHook
+  ] ++ lib.optional (!stdenv.isLinux) qtinstaller;
+
+  buildInputs = [
+    qtbase
+    qmltermwidget
+    qtdeclarative
+  ] ++ lib.optionals stdenv.isDarwin [ CoreFoundation Security ];
+  # FIXME: building on Darwin currently fails
+  # due to missing debug symbols for Qt,
+  # this should be fixable once darwin.apple_sdk >= 10.13
+  # See https://bugreports.qt.io/browse/QTBUG-76777
+
+  # Not using buildGoModule's buildPhase:
+  # gui/build.sh will build Go modules into lib/libgoshim.a
+  buildPhase = ''
+    runHook preBuild
+
+    make gen_providers_json
+    make generate
+    # Remove timestamps in comments
+    sed -i -e '/^\/\//d' pkg/config/version/version.go
+
+    # Not using -j$NIX_BUILD_CORES because the Makefile's rules
+    # are not thread-safe: lib/libgoshim.h is used before being built.
+    make build
+
+    runHook postBuild
+  '';
+
+  postInstall = ''
+    install -m 755 -D -t $out/bin build/qt/release/${provider}-vpn
+
+    VERSION=${version} VENDOR_PATH=providers branding/scripts/generate-debian branding/templates/debian/data.json
+    (cd branding/templates/debian && ${python3Packages.python}/bin/python3 generate.py)
+    install -m 444 -D branding/templates/debian/app.desktop $out/share/applications/${provider}-vpn.desktop
+  '' + lib.optionalString stdenv.isLinux ''
+    install -m 444 -D -t $out/share/polkit-1/actions ${bitmask-root}/share/polkit-1/actions/se.leap.bitmask.policy
+  '';
+
+  # Some tests need access to the Internet:
+  # Post "https://api.black.riseup.net/3/cert": dial tcp: lookup api.black.riseup.net on [::1]:53: read udp [::1]:56553->[::1]:53: read: connection refused
+  doCheck = false;
+
+  passthru = { inherit bitmask-root; };
+
+  meta = {
+    description = "Generic VPN client by LEAP";
+    longDescription = ''
+      Bitmask, by LEAP (LEAP Encryption Access Project),
+      is an application to provide easy and secure encrypted communication
+      with a VPN (Virtual Private Network). It allows you to select from
+      a variety of trusted service provider all from one app.
+      Current providers include Riseup Networks
+      and The Calyx Institute, where the former is default.
+      The <literal>${provider}-vpn</literal> executable should appear
+      in your desktop manager's XDG menu or could be launch in a terminal
+      to get an execution log. A new icon should then appear in your systray
+      to control the VPN and configure some options.
+    '';
+    homepage = "https://bitmask.net";
+    license = lib.licenses.gpl3Plus;
+    maintainers = with lib.maintainers; [ julm ];
+    # darwin requires apple_sdk >= 10.13
+    platforms = lib.platforms.linux;
+  };
+}

--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -3949,6 +3949,11 @@ with pkgs;
 
   bluetooth_battery = python3Packages.callPackage ../applications/misc/bluetooth_battery { };
 
+  calyx-vpn = libsForQt5.callPackage ../tools/networking/bitmask-vpn {
+    provider = "calyx";
+    inherit (darwin.apple_sdk.frameworks) CoreFoundation Security;
+  };
+
   code-browser-qt = libsForQt5.callPackage ../applications/editors/code-browser { withQt = true;
                                                                                 };
   code-browser-gtk = callPackage ../applications/editors/code-browser { withGtk = true;
@@ -8391,6 +8396,11 @@ with pkgs;
 
   rig = callPackage ../tools/misc/rig {
     stdenv = gccStdenv;
+  };
+
+  riseup-vpn = libsForQt5.callPackage ../tools/networking/bitmask-vpn {
+    provider = "riseup";
+    inherit (darwin.apple_sdk.frameworks) CoreFoundation Security;
   };
 
   rocket = libsForQt5.callPackage ../tools/graphics/rocket { };

